flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java

```diff
@@ -387,14 +403,25 @@ public int getChainLength() {
 
 		protected final StreamStatusProvider streamStatusProvider;
 
-		public ChainingOutput(OneInputStreamOperator<T, ?> operator, StreamStatusProvider streamStatusProvider) {
+		protected final OutputTag<T> outputTag;
+
+		public ChainingOutput(
+				OneInputStreamOperator<T, ?> operator,
+				StreamStatusProvider streamStatusProvider,
+				OutputTag<T> outputTag) {
 			this.operator = operator;
 			this.numRecordsIn = ((OperatorMetricGroup) operator.getMetricGroup()).getIOMetricGroup().getNumRecordsInCounter();
 			this.streamStatusProvider = streamStatusProvider;
+			this.outputTag = outputTag;
 		}
 
 		@Override
```

Comment:
Again the two `collect()` methods have much identical code. We can put that common code in a separate private method and calls this instead of repeating the code.